src/modulation.jl

```diff
@@ -0,0 +1,194 @@
+abstract type Modulation end
+abstract type BOC <: Modulation end
+
+struct BOCsin{M <: Union{AbstractFloat, Integer}, N <: Union{AbstractFloat, Integer}} <: BOC
+    m::M
+    n::N
+    BOCsin(m, n) =
+        m >= 1 && n >= 1 ?
+        new{typeof(m), typeof(n)}(m, n) :
+        error("m and n must be >= 1")
+end
+
+struct BOCcos{M <: Union{AbstractFloat, Integer}, N <: Union{AbstractFloat, Integer}} <: BOC
+    m::M
+    n::N
+    BOCcos(m, n) =
+        m >= 1 && n >= 1 ?
+        new{typeof(m), typeof(n)}(m, n) :
+        error("m and n must be >= 1")
+end
+
+struct LOC <: Modulation end
+
+struct CBOC{B1 <: BOC, B2 <: BOC} <: BOC
+    boc1::B1
+    boc2::B2
+    boc1_power::Float32
+    CBOC(boc1, boc2, boc1_power) =
+        0 < boc1_power < 1 && boc1.n == boc2.n ?
+        new{typeof(boc1), typeof(boc2)}(boc1, boc2, boc1_power) :
+        error("Power of BOC1 must be between 0 and 1 and n of both BOCs must match")
+end
+
+get_code_type(system::T) where T <: AbstractGNSS = get_code_type(system, get_modulation(T))
+
+get_code_type(system::AbstractGNSS{<:AbstractMatrix{T}}, modulation) where T = T
+get_code_type(system::AbstractGNSS{<:AbstractMatrix{T}}, modulation::CBOC) where T = typeof(modulation.boc1_power)
```

Comment:
I think here we should use type promotion to be save, even if the type of the PRN matrix is more general than the type of the power ratio.
```suggestion
get_code_type(system::AbstractGNSS{<:AbstractMatrix{T}}, modulation::CBOC) where T = promote_type(T, typeof(modulation.boc1_power))
```